src/v2/client/client.ts

```diff
@@ -0,0 +1,206 @@
+import axios, { AxiosError, AxiosInstance, AxiosRequestConfig } from "axios";
+import { ClientCredentials } from "simple-oauth2";
+
+export type Request = {
+  path: string[];
+  /**
+   * Url Paramters
+   */
+  params?: Record<string, string | number | boolean>;
+  /**
+   * Request body will be serialized to json
+   */
+  body?: unknown;
+
+  headers?: Record<string, string | number | boolean>;
+
+  /**
+   * Retry the request a number of times while backing off exponentially
+   */
+  retry?: number;
+
+  /**
+   * in milliseconds
+   * @default 7000
+   */
+  timeout?: number;
+
+  /**
+   * Override the base Url for this request
+   */
+  baseUrl?: string;
+};
+
+export type ZohoResponse<TResponse> = TResponse & {
+  /**
+   * Zoho Inventory error code. This will be zero for a success response and non-zero in case of an error.
+   */
+  code: number;
+  /**
+   * Message for the invoked API.
+   */
+  message: string;
+
+  page_context?: {
+    page: number;
+    per_page: number;
+    has_more_page: boolean;
+  };
+};
+
+export class ZohoApiError extends Error {
+  url?: string;
+
+  code?: number;
+
+  constructor(err: AxiosError<{ code: number; message: string }>) {
+    super(err.response?.data.message ?? err.message);
+    this.url = err.request?.baseUrl + err.request?.path;
+
+    this.code = err.response?.data.code;
+  }
+}
+
+export type ZohoApiClientConfig = {
+  orgId: string;
+  headers: Record<string, string>;
+  baseUrl?: string;
+};
+
+export class ZohoApiClient {
+  private httpClient: AxiosInstance;
+
+  static readonly BASE_URL = {
+    inventory: "https://inventory.zoho.eu/api/v1",
+    books: "https://books.zoho.eu/api/v1",
+  };
+
+  private constructor(config: ZohoApiClientConfig) {
+    this.httpClient = axios.create({
+      baseURL: config.baseUrl ?? ZohoApiClient.BASE_URL.inventory,
+      headers: config.headers,
+      params: {
+        organization_id: config.orgId,
+      },
+      timeout: 30_000,
+    });
+  }
+
+  /**
+   * Create a zoho api instance from client id and secret
+   */
+  static async fromOAuth(config: {
+    orgId: string;
+    client: { id: string; secret: string };
+    baseUrl?: string;
+  }): Promise<ZohoApiClient> {
+    const clientCredentials = new ClientCredentials({
+      client: config.client,
+      auth: {
+        tokenHost: "https://accounts.zoho.eu",
+        tokenPath: "/oauth/v2/token",
+      },
+      options: {
+        authorizationMethod: "body",
+      },
+    });
+
+    const res = await clientCredentials.getToken({
+      scope: "ZohoInventory.FullAccess.all",
```

Comment:
@chronark we have to add here the zoho books scope, if we want to use the Books API. - We can start easily with "ZohoBooks.fullaccess.all"